spacy_experimental/span_boundary_detection/sbd_component.py

```diff
@@ -0,0 +1,301 @@
+            is_end = 0
+            if token.i in start_indices:
+                is_start = 1
+            if token.i in end_indices:
+                is_end = 1
+            reference_results.append((is_start, is_end))
+
+    return reference_results
+
+
+def get_predictions(docs) -> Floats2d:
+    """Create a prediction list of token start/end probabilities for evaluation"""
+    prediction_results = []
+    for doc in docs:
+        for token in doc:
+            prediction_results.append([token._.span_start, token._.span_end])
+    return prediction_results
+
+
+class SpanBoundaryDetection(TrainablePipe):
+    """Pipeline that learns start and end tokens of spans"""
+
+    def __init__(
+        self,
+        nlp: Language,
+        model: Model[List[Doc], Floats2d],
+        name: str = "sbd",
+        *,
+        threshold: float = 0.5,
+        scorer: Optional[Callable] = sbd_score,
+    ) -> None:
+        """Initialize the span boundary detector.
+        model (thinc.api.Model): The Thinc Model powering the pipeline component.
+        name (str): The component instance name, used to add entries to the
+            losses during training.
+        threshold (float): Minimum probability to consider a prediction
+            positive.
+        scorer (Optional[Callable]): The scoring method.
+        """
+        self.vocab = nlp
+        self.cfg = {
+            "threshold": threshold,
+        }
+        self.model = model
+        self.name = name
+        self.scorer = scorer
+        Token.set_extension("span_start", default=0, force=True)
+        Token.set_extension("span_end", default=0, force=True)
+
+    def predict(self, docs: Iterable[Doc]):
+        """Apply the pipeline's model to a batch of docs, without modifying them.
+        docs (Iterable[Doc]): The documents to predict.
+        RETURNS: The models prediction for each document.
+        """
+        scores = self.model.predict(docs)
+        return scores
+
+    def set_annotations(self, docs: Iterable[Doc], scores: Floats2d) -> None:
+        """Modify a batch of Doc objects, using pre-computed scores.
+        docs (Iterable[Doc]): The documents to modify.
+        scores: The scores to set, produced by SpanCategorizer.predict.
+        """
+        lengths = [len(doc) for doc in docs]
+
+        offset = 0
+        scores_per_doc = []
+        for length in lengths:
+            scores_per_doc.append(scores[offset : offset + length])
+            offset += length
+
+        for doc, score_doc in zip(docs, scores_per_doc):
+            for token, score_token in zip(doc, score_doc):
+
+                if score_token[0] > self.cfg["threshold"]:
+                    score_token[0] = 1
+                else:
+                    score_token[0] = 0
+
+                if score_token[1] > self.cfg["threshold"]:
+                    score_token[1] = 1
+                else:
+                    score_token[1] = 0
+
+                token._.span_start = score_token[0]
```

Comment:
If neither of these are true then `token.span_boe` would be `0` and otherwise `1` or `2`. I don't think its crucial, maybe later for consistency.